accepted/2020/mono-convergence/platform-specific-httpclient.md

```diff
@@ -0,0 +1,452 @@
+```c#
+// System.Net.Http.HandlerFactory.dll
+public static class HttpHandlerFactory
+{
+  // It could be of object type not to take dependency on System.Net.Http
+  public HttpClientHandler Create () => throw new NotImplementedException();
+}
+```
+
+```c#
+// System.Net.Http.dll
+public HttpClientHandler ()
+{
+  platformInstance = HttpHandlerFactory.Create();
+}
+```
+
+**Disadvantages**
+
+* Single method assembly would be added to the shared framework
+* A circular dependency between System.Net.Http.HandlerFactory.dll and System.Net.Http.dll will exist if we use HttpClientHandler as the “common” type.
+
+**Proposal 2d: Add SDK reference assemblies with factory contract to dotnet/runtime**
+
+This is what Mono does today. The reference assemblies which describe the factory method contract will be available during dotnet/runtime build and used for each RID specific build of System.Net.Http.dll.
+```c#
+// System.Net.Http.dll
+public HttpClientHandler ()
+{
+#if iOS
+  platformInstance = Xamarin.iOS.HttpHandleFactory.Create();
+#else
+...
+#endif
+}
+```
+
+### Proposal #3: Move HttpClientHandler to a new assembly
+
+The assembly called System.Net.Http.Private.dll would be introduced and would include only implementation of HttpClientHandler. The implementation then would be fully platform specific (per RID) and **for Xamarin like SDKs coming from their optional SDK pack**. For the desktop version version we could just leave HttpClientHandler in implementation version of System.Net.Http.dll.
+
+**Located in dotnet/runtime repo**
+```c#
+// System.Net.Http.dll
+
+#if iOS || android || tvOS || wasm
+// The reference for the type comes from System.Net.Http.Private.dll
+[assembly: TypeForwardTo (typeof (System.Net.Http.HttpClientHandler)) 
+#endif
+```
+
+**Located in xamarin/xamarin-macios repo**
+```c#
+// Sample implementation in System.Net.Http.Private.dll build as part of Xamarin.iOS
+public class HttpClientHandler
+{
+  public HttpClientHandler ()
+  {
+  }
+
+  public virtual bool SupportsProxy => nativecall ("has-proxy") as bool;
+
+  public int MaxAutomaticRedirections {
+    get => 100;
+    set => throw PlatformNotSupportedException();
+  }
+
+  // The rest of API contract follows
+```
+
+**Disadvantages**
+
+* Single type assembly is against our design practices
+* A circular dependency between System.Net.Http.Private.dll and System.Net.Http.dll might exist in optional components
+* API contract would live in different assembly than implementation
+* Reliance on more dotnet publish to figure out right implementation assembly
+* Adding public member to HttpClientHandler would be a breaking change
+* HttpClientHandler acts as a wrapper in most SDKs and the code would have to be duplicated
+* Platform specific implementation would need to write their own copy of proxy logic
```

Comment:
That's a breaking change for SDK since they already have the handlers (plural for XI/XM).
The solution must allow the selection of `CFNetworkHander` and `NSUrlSessionHandler` (today).
We might be able to kill `CFNetworkHander` (at some point, e.g. Apple dropping iOS7 support) but Apple might decide to _gift us_ something new (forcing us to support more than one handler again).

Reply:
The proposal allows any logic to exist in SDKs so they could create any handler